### ISSUE TITLE
`brew install` should display download size

### ISSUE THREAD

**Issue by osalbahr** [Homebrew/brew#18373]
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

I want to know the total download/install size when installing a package.

### What is the motivation for the feature?

I am using a limited storage system (16 GB). Also boredom and I want to contribute to Homebrew again.

### How will the feature be relevant to at least 90% of Homebrew users?

Sort of? It would be nice. But probably not needed. Storage is cheap. It will only affect people with pay-per-use data plans.

### What alternatives to the feature have been considered?

`brew deps` then somehow find tarball size for each package. This is good enough for my use.

Example (only first step):
```console
$ echo $(brew deps fastfetch)
binutils gcc gmp isl libmpc lz4 mpfr xz zlib zstd
```

Then a bash for-each, since formula names are guaranteed to not have spaces (I think).

But would be nice if it's integrated into Homebrew under an undocumented flag. I would appreciate pointers on how to start this, such as existing open issues/discussions. Thx!

More info:

```console
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: No developer tools installed.
Install Clang or run `brew install gcc`.
$ brew config
HOMEBREW_VERSION: 4.3.23-53-g29c22e0
ORIGIN: https://github.com/Homebrew/brew
HEAD: 29c22e0ab3fd9826cefd4b14c143266dea3bc391
Last commit: 2 days ago
Core tap JSON: 22 Sep 11:36 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: /usr/bin/nano
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.4 => /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.4_1/bin/ruby
CPU: octa-core 64-bit sandybridge
Clang: N/A
Git: 2.46.1 => /bin/git
Curl: 8.6.0 => /bin/curl
Kernel: Linux 6.10.10-200.fc40.x86_64 x86_64 GNU/Linux
OS: Fedora release 40 (Forty)
Host glibc: 2.39
/usr/bin/gcc: N/A
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: N/A
xorg: N/A
```

**Comment by osalbahr** [comment]
Here is a minimal way to find the total tarball sizes:

```bash
for P in $(brew deps fastfetch)
do
    brew cat "$P" | \
    grep -m1 url | \
    awk '{print $2}' | \
    xargs wget -qO- | \
    wc -c
done
```

Note: tapping and downloading is required by the above for loop which (sort of) defeats the points, but this is merely a proof of concept. I think there should be a "size" field right below the `sha256` value in a formula. Also theoretically a "security" measure to check both the shasum and size, but that's a later decision.

**Comment by carlocab** [comment]
> I think there should be a "size" field right below the `sha256` value in a formula.

That's a bit tricky since the bottle contains a copy of the formula, so changing the formula changes the size of the bottle too.

That said, I think the bottle size (both compresses and uncompressed) is available in the annotations in the manifest, so in principle this should be straightforward to recover with `curl` and `jq`. I don't think it contains source tarball size, but most users probably don't really care about this (because they install formulae from bottles).

Or, in a pinch, if you only care about compressed size:
```
brew fetch foo
wc -c "$(brew --cache foo)" # bottle size

brew fetch --build-from-source foo
wc -c "$(brew --cache --build-from-source foo)" # source tarball size
```
This makes it so that you don't need to download `foo` all over again if you want to install it or build it from source.

**Comment by osalbahr** [comment]
> That said, I think the bottle size (both compresses and uncompressed) is available in the annotations in the manifest, so in principle this should be straightforward to recover with `curl` and `jq`. I don't think it contains source tarball size, but most users probably don't really care about this (because they install formulae from bottles).

I agree. The size of bottle is what I was looking for rather than the tarballs.

> That said, I think the bottle size (both compresses and uncompressed) is available in the annotations in the manifest

That sounds like what I was looking for. Is there perhaps some sort of API that lets me "fetch" only the size, without first downloading the bottle?

And `wc -c "$(brew --cache foo)"` does seem to print the intended "Download Size" of a specific formula. But to my understanding I have to have the bottle cached locally first.

**Comment by carlocab** [comment]
> That sounds like what I was looking for. Is there perhaps some sort of API that lets me "fetch" only the size, without first downloading the bottle?

Yes. The manifest can be fetched separately from the bottle. (In fact, `brew` downloads the manifest first before downloading the bottle.)

> And `wc -c "$(brew --cache foo)"` does seem to print the intended "Download Size" of a specific formula. But to my understanding I have to have the bottle cached locally first.

Yes, that's why you need to `brew fetch` first.

**Comment by osalbahr** [comment]
Where can I read more about getting only the manifest? I couldn't find it documented in https://docs.brew.sh.

**Comment by carlocab** [comment]
It's not documented, but you can see what `brew` does to fetch it by doing `brew fetch --verbose`:
```
❯ brew fetch --force --verbose libuv
==> Downloading https://ghcr.io/v2/homebrew/core/libuv/manifests/1.48.0
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.3.23-54-gde9848d\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 15.0\)\ curl/8.7.1 --header Accept-Language:\ en --fail --retry 3 --header Accept:\ application/vnd.oci.image.index.v1+json --header Authorization:\ Bearer\ QQ== --remote-time --output /Users/carlocab/Library/Caches/Homebrew/downloads/8ee1d27fb604f55e3c4415b96e34dc9c7f557996832c372d984c8162b29a4ad2--libuv-1.48.0.bottle_manifest.json.incomplete --location https://ghcr.io/v2/homebrew/core/libuv/manifests/1.48.0
```

**Comment by osalbahr** [comment]
Now I'm thinking I perhaps should first work on a `brew fetch --manifest-only foo`. What do you think?

**Comment by carlocab** [comment]
Seems to have a relatively niche/obscure use-case at the moment, so it's probably not a good fit.

**Comment by osalbahr** [comment]
Ok. I'll see what I can do. Thx!

**Comment by umnikos** [comment]
One time I tried to install a calculator with `brew`. It took a weirdly long amount of time, and then I saw this:

`Installing libqalculate dependency: xorg-server`

I think I'd personally benefit from brew calculating some kind of size for the operation (download size should be easy) and then giving the user a yes/no prompt if that size is large.

**Comment by osalbahr** [comment]
I am no longer interested in solely working on a PR. But happy to help. @umnikos if you (or someone else) are interested, lmk, and I will re-open the issue. Happy to help you debug too, just @ me.

I will also consider working on this given enough thumbs up. I honestly do not really care and thought it was more trivial than it actually is. There isn't an API for my purpose. But if there is enough interested, it might be fun to hack up.

**Comment by MikeMcQuaid** [comment]
> brew calculating some kind of size for the operation (download size should be easy)

@umnikos if you read the whole thread you'll see why this is not easy. If you disagree: we'd welcome said easy PR 😉 

> t turned out to be more complicated than I expected as there is no API to fetch the download/install size.

@osalbahr as mentioned by @carlocab above: said API is "download the manifest first and use that to figure out the size of the formula".

The longer-term solution here would be to consider if we could e.g. download all these manifests daily and provide a JSON file of the rough sizes that could be incorporated into the formulae.brew.sh JSON API. I'd imagine ~24h outdated information here would be more useful than none at all.

**Comment by osalbahr** [comment]
> @osalbahr as mentioned by @carlocab above: said API is "download the manifest first and use that to figure out the size of the formula".

I tried to work on that, but the path for going from `libuv` for example and the long one-liner seemed intimidating. My approach is usually to hack something up in a language I know, like `bash`, then figure out how to put that in a language I am not familiar with, like Ruby. However I think I might need to directly go at this issue in Ruby directly.

Could you point me at the Ruby function that installs the manifest? It is difficult to figure that out since I couldn't find the manifest in the technical docs.

**Comment by apainintheneck** [comment]
It's not exactly what you're looking for but @cho-m opened a proof of concept PR (https://github.com/Homebrew/brew/pull/18172) a couple weeks ago to add formula size information to `brew info`. The size information gets pulled from the manifest which seems similar to what you all are talking about here. Of course, it doesn't incorporate that information into `brew install` but maybe there's some logic that could be shared from that PR if/when it gets merged in. At the very least it seems to be tangentially related to this discussion.

**Comment by cho-m** [comment]
I did have another relate P.O.C. of creating a summed size. Mainly was experimenting with it as part of `--dry-run` (and also hacking that into some user prompting feature), e.g. 
```diff
--- a/Library/Homebrew/install.rb
+++ b/Library/Homebrew/install.rb
@@ -284,17 +285,27 @@ module Homebrew
           end
         end

-        if dry_run
+        if dry_run || prompt
           if (formulae_name_to_install = formulae_to_install.map(&:name))
             ohai "Would install #{Utils.pluralize("formula", formulae_name_to_install.count,
                                                   plural: "e", include_count: true)}:"
             puts formulae_name_to_install.join(" ")

+            formulae = Set.new
             formula_installers.each do |fi|
               print_dry_run_dependencies(fi.formula, fi.compute_dependencies, &:name)
+              formulae.add(fi.formula)
+              formulae.merge(fi.compute_dependencies.flatten.map(&:to_formula))
             end
+            installed_sizes = formulae.map { |f| f.bottle.installed_size }
+            ohai "Total size: #{disk_usage_readable(installed_sizes.sum)}" if installed_sizes.all?(&:positive?)
+          end
+          return if dry_run
```

---

Anyway, will try to get around to my original PR first. Hopefully will have some more time.

**Comment by osalbahr** [comment]
@cho-m awesome! Let me know if you'd like some help. We're not doing the exact same, but we can build off of a common ground. I like @MikeMcQuaid's idea of adding bottle sizes in JSON API.

**Comment by osalbahr** [comment]
Oh, [`disk_usage_readable`](https://github.com/Homebrew/brew/blob/bd3c7f80530d21e791ccc01b02234b299941dd29/Library/Homebrew/extend/kernel.rb#L450-L471) looks like a cool function. But I first need to figure out how to get the `size_in_bytes`.

@cho-m btw, how did you get `installed_sizes.sum`? I can't find it in the main fork or your fork (or global GitHub search). Perhaps you haven't pushed it yet or it's in a private fork.

**Comment by MikeMcQuaid** [comment]
@cho-m @osalbahr Note: if/when we're adding more information to the API here: it may be nice to also include `sh.brew.path_exec_files` from the tab/manifest, too; that'll make https://github.com/homebrew/homebrew-command-not-found much nicer.

**Comment by osalbahr** [comment]
Since I am just learning Ruby, I think `brew info` is an easier milestone.

@cho-m what do you think?

**Comment by MikeMcQuaid** [comment]
> Since I am just learning Ruby, I think `brew info` is an easier milestone.

@osalbahr Feel free to start with just that.

**Comment by osalbahr** [comment]
> > Since I am just learning Ruby, I think `brew info` is an easier milestone.
> 
> @osalbahr Feel free to start with just that.

I appreciate the green light. It lets me know that I am moving in the right direction. Though I don't mind if someone beats me to it.

I [created a repo learn Ruby](https://github.com/osalbahr/learn-ruby). I will come back to adding the `brew info` feature later. Or, I might script a working solution in Bash first. I will see.

**Comment by osalbahr** [comment]
I am trying to add a line under `==> Analytics` that says `bottle-size: 0` as a placeholder. However I can't figure out where the lines under Analytics come from. Could you help by pointing at which Ruby file is responsible for the list?

I looked at `info.rb` and I am not sure where the delegation happens.

```console
$ brew info fastfetch | tail
==> Options
--HEAD
        Install HEAD version
==> Caveats
Bash completion has been installed to:
  /home/linuxbrew/.linuxbrew/etc/bash_completion.d
==> Analytics
install: 13,275 (30 days), 31,864 (90 days), 65,133 (365 days)
install-on-request: 13,275 (30 days), 31,864 (90 days), 65,133 (365 days)
build-error: 1 (30 days)
```

**Comment by ZhongRuoyu** [comment]
That happens here:

https://github.com/Homebrew/brew/blob/ad9ea1cb04898020188b551de84b3d350fa28009/Library/Homebrew/utils/analytics.rb#L327

Invoked from here in `info.rb`:

https://github.com/Homebrew/brew/blob/ad9ea1cb04898020188b551de84b3d350fa28009/Library/Homebrew/cmd/info.rb#L360

**Comment by osalbahr** [comment]
@ZhongRuoyu thx!

I figured out adding the placeholder. The next step would be figuring out how to get it from (or add it to) the JSON API.

```console
$ brew info fastfetch | tail -2
==> Size
Bottle size: 0
```

**Comment by carlocab** [comment]
Is there anything left to do here? I think this has already been implemented. For example:
```
❯ brew info hello
==> hello: stable 2.12.1 (bottled)
[snip]
Not installed
Bottle Size: 51KB
Installed Size: 178.3KB
[snip]
```

**Comment by osalbahr** [comment]
@carlocab what branch are you on? I don't see the Size lines on `master`.

**Comment by carlocab** [comment]
```
❯ brew --version
Homebrew 4.4.11-27-g9783ab0
Homebrew/homebrew-core (git revision ebac45301ce; last commit 2024-12-12)
Homebrew/homebrew-cask (git revision 8b7c694292a; last commit 2024-12-12)
```

**Comment by osalbahr** [comment]
Not sure what is happening.

```console
$ brew --version
Homebrew 4.4.11-27-g9783ab0
Homebrew/homebrew-core (git revision 249e9860b75; last commit 2024-12-12)
$ brew info hello
==> hello: stable 2.12.1
Program providing model for GNU coding standards and practices
https://www.gnu.org/software/hello/
Conflicts with:
  perkeep (because both install `hello` binaries)
Not installed
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/h/hello.rb
License: GPL-3.0-or-later
==> Analytics
install: 11,230 (30 days), 14,692 (90 days), 33,990 (365 days)
install-on-request: 11,230 (30 days), 14,692 (90 days), 33,988 (365 days)
build-error: 18 (30 days)
```

I double-checked that I don't have `git` changes:
```console
$ cd $(brew --repo)
$ git branch
  info-size
* master
  stable
$ git status
On branch master
nothing to commit, working tree clean
```

**Comment by carlocab** [comment]
What's the output of `brew config` and `brew doctor`?

**Comment by osalbahr** [comment]
```console
% brew config
HOMEBREW_VERSION: 4.4.11-27-g9783ab0
ORIGIN: https://github.com/Homebrew/brew
HEAD: 9783ab062cfcdc5ce776bfe5d163be3ee5fc7840
Last commit: 12 hours ago
Branch: master
Core tap HEAD: a39a029f686500517cc653fb5436d5a9ca50e52a
Core tap last commit: 53 minutes ago
Core tap JSON: 12 Dec 19:27 UTC
Core cask tap HEAD: 42778f1712c375cd3822b8223ade28c2545922b6
Core cask tap last commit: 55 minutes ago
Core cask tap JSON: 12 Dec 19:27 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.6 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.6/bin/ruby
CPU: 10-core 64-bit arm_blizzard_avalanche
Clang: 16.0.0 build 1600
Git: 2.39.5 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.1.1-arm64
CLT: 16.1.0.0.1.1729049160
Xcode: N/A
Rosetta 2: false
```

```console
% brew doctor
Your system is ready to brew.
```

**Comment by carlocab** [comment]
Ah, I see what's going on. I've previously run `brew info --fetch-manifest hello`, so the size data is already cached.

Try doing `brew info --fetch-manifest hello` first. This will show the size information. Future invocations of `brew info hello` should then use the cached size data without needing `--fetch-manifest`.

**Comment by cho-m** [comment]
There is still a goal of adding data to JSON API, though whoever works on this needs to make sure on any performance impact, e.g.
* JSON API generation time. Simply fetching all 7000+ formulae manifests from GitHub is slow.
* JSON size. API performance is related to JSON size so should benchmark impact. If noticeable, then need to figure out how to improve performance first. Related: #16410

**Comment by osalbahr** [comment]
> Try doing `brew info --fetch-manifest hello` first. This will show the size information. Future invocations of `brew info hello` should then use the cached size data without needing `--fetch-manifest`.

Yup, that was it.

Is this intended behavior?

**Comment by cho-m** [comment]
> Is this intended behavior?

Yes as fetching manifest adds a short but noticeable impact. Could consider adding an environment variable to make it default for users who are aware and willing to accept the extra network request.

Ideally want to default to fastest/offline-capable `brew info` output (though default still requires fetching installation analytics - https://github.com/Homebrew/brew/pull/17960)

**Comment by tyuwags** [comment]
do you think a format as a pacman installation style could be some idea to present dependency, download size and installation choices ?
sudo pacman -S ruby
resolving dependencies...
looking for conflicting packages...
warning: dependency cycle detected:
warning: rubygems will be installed before its ruby dependency

Packages (2) rubygems-3.5.16-1  ruby-3.3.5-2

Total Download Size:    4.58 MiB
Total Installed Size:  18.83 MiB

:: Proceed with installation? [Y/n]

**Comment by osalbahr** [comment]
> do you think a format as a pacman installation style could be some idea to present dependency, download size and installation choices ?

Yes. Although I think download and install size would be the same, so showing only the download size is sufficient.

**Comment by tyuwags** [comment]
okay so both download and install size exist so maybe I will put both. the problem is that for no command, brew asks for more input of the user, other than the password for specific software. I don't know then if its a good idea to implement this or not and to be honest I don't know how to do it securely in ruby without any overflow and other problems

**Comment by MikeMcQuaid** [comment]
> the problem is that for no command, brew asks for more input of the user, other than the password for specific software.

Yup, it'd need to be `brew install --ask` and/or a `HOMEBREW_ASK` environment variable.

**Comment by tyuwags** [comment]
How could I get some reviews before creating a pull request? Or should I do it directly ?

**Comment by gromgit** [comment]
You'd have to push a PR before anyone can review your code.
